### ISSUE TITLE
Community contribution: Adding GGUF support for more architectures

### ISSUE THREAD

**Issue by SunMarc** [huggingface/transformers#33260]
### Feature request

Recently, we have added the ability to load `gguf` files within [transformers](https://huggingface.co/docs/hub/en/gguf).

<img src="https://github.com/user-attachments/assets/61df6455-6016-449e-a37f-9dfc7f918902" width="600">

The goal was to offer the possibility to users to further train/fine-tune their gguf models. 
<details>
<summary>See Workflow</summary>
1) Load gguf file in transformers: we dequantize the weights to fp32, then we load the weights to be used with PyTorch.

```py 
from transformers import AutoTokenizer, AutoModelForCausalLM

model_id = "TheBloke/TinyLlama-1.1B-Chat-v1.0-GGUF"
filename = "tinyllama-1.1b-chat-v1.0.Q6_K.gguf"

tokenizer = AutoTokenizer.from_pretrained(model_id, gguf_file=filename)
model = AutoModelForCausalLM.from_pretrained(model_id, gguf_file=filename)
```
2) train/finetune

3) Convert the model back to gguf to use in the ggml ecosystem using [convert_hf_to_gguf](https://github.com/ggerganov/llama.cpp/blob/master/convert_hf_to_gguf.py) script or using [gguf-my-repo](https://huggingface.co/spaces/ggml-org/gguf-my-repo) space if you pushed your model on the hub :
```py
tokenizer.save_pretrained('directory')
model.save_pretrained('directory')

!python ${path_to_llama_cpp}/convert-hf-to-gguf.py ${directory}
```
</details>

Let's try to add GGUF support for more architectures! Currently supported architectures are

- [x] Llama
- [x] Mistral
- [x] Qwen2

It would be great to add the support for more architectures such as

- [ ] Phi3 https://github.com/huggingface/transformers/pull/31844
- [ ] Qwen2Moe
- [ ] Gemma2

... and many more (feel free to suggest more architectures !)

Adding this feature would require to follow the same protocol as in this [PR](https://github.com/huggingface/transformers/pull/31175/files) : 
1) Update `GGUF_TENSOR_MAPPING` and `GGUF_CONFIG_MAPPING` in order to map the tensor/config of the gguf file to the one on transformers. 
2) Create a `GGUFXXXConverter(XXXConverter)` class to convert the gguf tokenizer to a transformers one. 
3) Write tests

If you are interested to take up the challenge, comment below with the architecture name you want to integrate and open a PR! 

Once you open a PR, feel free to ping @SunMarc @LysandreJik @ArthurZucker for a review ! 

### Motivation

Support for more gguf models

### Your contribution

Reviewing PRs and possibly adding the support for more models

**Comment by KingNish24** [comment]
@SunMarc I want to take Gemma2

**Comment by junejae** [comment]
May I suggest & take T5? Seems [GGUF version of T5 encoder](https://huggingface.co/city96/t5-v1_1-xxl-encoder-gguf) is highly used for getting along with FLUX.

**Comment by 010kim** [comment]
@SunMarc Hello! Unless someone else is working on this model already, may I take MiniCPM-V?

**Comment by SunMarc** [comment]
> @SunMarc May I suggest & take T5? Seems [GGUF version of T5 encoder](https://huggingface.co/city96/t5-v1_1-xxl-encoder-gguf) is highly used for getting along with FLUX.

Added @junejae !

> @SunMarc Hello! Unless someone else is working on this model already, may I take MiniCPM-V?

Hi @010kim, thanks for the interest ! MiniCPM-V model relies on `trust_remote_code=True`, so I don't think we can add this model for now with gguf support. We don't want to have code in transformers that relies on modeling files that are on the hub. I will think about extending `trust_remote_code=True` to gguf support, so that the author of the model can add it himself !

**Comment by 010kim** [comment]
> Hi @010kim, thanks for the interest ! MiniCPM-V model relies on `trust_remote_code=True`, so I don't think we can add this model for now with gguf support. We don't want to have code in transformers that relies on modeling files that are on the hub. I will think about extending `trust_remote_code=True` to gguf support, so that the author of the model can add it himself !

@SunMarc Thank you so much for your response. It also makes sense the author should work on it. What about Cohere? Can I take it?

**Comment by jungnerd** [comment]
Hi @SunMarc 👋🏻
May I work with CLIP model if nobody is working on it?

**Comment by SunMarc** [comment]
Hey @jungnerd ! The model you choose needs to be in the conversion script from hf to gguf. See the script [here](https://github.com/ggerganov/llama.cpp/blob/master/convert_hf_to_gguf.py)

**Comment by g-prz** [comment]
Hey @SunMarc 🙋‍♂️ 
I'd like to try my chance to contribute to this issue, can I take Falcon? 🦅

**Comment by VladOS95-cyber** [comment]
Hi @SunMarc, I take bloom if nobody is working on it

**Comment by fabxoe** [comment]
Hi @SunMarc, I’ll handle the work related to Codestrall :)

**Comment by jungnerd** [comment]
> Hey @jungnerd ! The model you choose needs to be in the conversion script from hf to gguf. See the script [here](https://github.com/ggerganov/llama.cpp/blob/master/convert_hf_to_gguf.py)

There is conversion script for clip model([clip.cpp](https://github.com/monatis/clip.cpp/blob/main/models/convert_hf_to_gguf.py)). Can I use this to contribute?

**Comment by cjfghk5697** [comment]
Hi @SunMarc,
I'm interested in this issue. Would it be okay if I worked on the BLIP model?

**Comment by cjfghk5697** [comment]
> Hi @SunMarc, I'm interested in this issue. Would it be okay if I worked on the BLIP model?

Hi @SunMarc, I'd like to work on the BLIP model, but after researching, I found that it might be challenging due to the Vision model structure. Would it be alright if I switched to working on the Smol model instead?

**Comment by g-prz** [comment]
Hey @SunMarc 🤗 
Gonna continue with granite 🪨

**Comment by cjfghk5697** [comment]
@SunMarc 
I checked the Smol model and confirmed that it's already functioning well without needing any further work. In the issue, I mentioned that supporting the Smol model would be beneficial, but is there any specific work required?

If not, I’ll proceed with switching to the dbrx model.

**Comment by SunMarc** [comment]
> @SunMarc
> I checked the Smol model and confirmed that it's already functioning well without needing any further work. In the issue mentioned that supporting the Smol model would be beneficial, but is there any specific work required?
> 
> If not, I’ll proceed with switching to the dbrx model.

Oh indeed, this is because it is a llama architecture.

**Comment by VladOS95-cyber** [comment]
Hi @SunMarc! I am going to start working on MPT model

**Comment by yijun-lee** [comment]
Is any work being done on the Gemma2? If not, I would like to proceed with it!
@SunMarc @KingNish24

**Comment by VladOS95-cyber** [comment]
Hi @SunMarc! I suppose GPT2 gguf is not supported yet, if this is a case, I'll take it

**Comment by fabxoe** [comment]
> Hi @SunMarc, I'd like to handle the work related to Codestrall :)

Codestrall's tokenizer was just llama2.
It looks like I don't need to handle codes of Codestrall

**Comment by 010kim** [comment]
> @SunMarc Thank you so much for your response. It also makes sense the author should work on it. What about Cohere? Can I take it?

I went through the codes, and i was able to to load Cohere gguf model, but could not load the tokenizer. This is because Cohere slow tokenizer is not implemented in HuggingFace. (Only FastTokenizer is available for Cohere) Is there a way around to fix this? @SunMarc

**Comment by VladOS95-cyber** [comment]
Hey @SunMarc! I'll take StarCoder2 as a next model

**Comment by VladOS95-cyber** [comment]
Hey @SunMarc! I"ll take Starcoder2 as next model

**Comment by VladOS95-cyber** [comment]
Hi @SunMarc! I am going to start working on Mamba

**Comment by farrosalferro** [comment]
Are you still working on Gemma2? @yijun-lee @KingNish24 ? If not, is it possible for me to try working on it? Thank you!

**Comment by farrosalferro** [comment]
Could you please kindly check my PR @SunMarc? Thank you 
[Add Nemotron GGUF Loading Support](https://github.com/huggingface/transformers/pull/34725)

**Comment by mreraser** [comment]
Hello everyone! I truly admire the amazing work you’re all doing. If no one has started working on [RoBERTa](https://huggingface.co/docs/transformers/main/en/model_doc/roberta) yet, may I kindly propose taking on that task myself?

**Comment by zinccat** [comment]
anyone working on supporting DeepSeek V3?

**Comment by SunMarc** [comment]
Deekpseek V3 is not supported yet in transformers but it will be soon with this [PR](https://github.com/huggingface/transformers/pull/35926#issuecomment-2636398699)

**Comment by zinccat** [comment]
Thanks, it should be quite similar to DeepSeekV2. I have made some progress on it, though mainly on another direction. https://github.com/zinccat/gguf_offload

**Comment by Mett-2002** [comment]
But it says Architecture qwen2 not supported
the model that I'm using:
gte-Qwen2-7B-instruct-Q4_K_M.gguf

**Comment by StefMa** [comment]
Can we also get support for gemma3 please.
Who is able to do that?
Maybe @kinshuk1207 because you also added support for the original gemma in https://github.com/huggingface/transformers/pull/35887?

**Comment by huzhekun** [comment]
Can we also get support for cohere2 as well? or are there issue preventing that from happening?

**Comment by SunMarc** [comment]
Which model in particular ?  Please share the model card @huzhekun

**Comment by huzhekun** [comment]
I am specifically personally looking at this model @SunMarc , any other info I should provide?
https://huggingface.co/bartowski/CohereForAI_c4ai-command-a-03-2025-GGUF

**Comment by SunMarc** [comment]
No thanks ! I'll add it above as a suggested architecture to add support for

**Comment by jhuntbach-bc** [comment]
Would love to get llama4 support
```
GGUF model with architecture llama4 is not supported yet
```

**Comment by SunMarc** [comment]
llama4 would be really nice to have indeed ! cc @Isotr0py

**Comment by Lori10** [comment]
Is llama4 supported ?

**Comment by tanuj-rai** [comment]
Hi @SunMarc, can I add support to the Phi4 multimodel?

**Comment by kunibald413** [comment]
qwen2vl gguf support

please support this gguf model: 

https://huggingface.co/mradermacher/SkyCaptioner-V1-GGUF

**Comment by sdhserrvesrhywe** [comment]
https://huggingface.co/city96/umt5-xxl-encoder-gguf please add gguf support to this model. @junejae does your work support his model？

**Comment by Kim-Ju-won** [comment]
Hi @SunMarc 

Thank you for all your work on expanding GGUF support in Transformers — it’s been great to see the progress! 

I was wondering if GGUF support could be extended to the **NAVER HyperCLOVAX models** (e.g. SEED-Text 0.5B/1.5B, SEED-Vision-3B, SEED-Think-14B).  
Some smaller SEED models already run well via community GGUF conversions, and official support would make workflows much smoother, especially for the reasoning and multimodal variants.

If this sounds valuable, I’d be glad to help or draft an initial PR.

**Comment by deepayansanyal93** [comment]
Hi @SunMarc, it looks like GGUF support for Llama 4 may be incomplete at the moment. I’ve tried loading some Llama 4 GGUF checkpoints and ran into issues.

I’d be happy to take a look and work on adding support for Llama4 here if this is still an open area.

**Comment by garybadwal** [comment]
Hi @SunMarc,
Can I work on Llama4 from the checklist? I have an implementation ready for this, please let me know if i can so i can push it and raise a PR for that.

**Comment by 0xpinara** [comment]
Hi @SunMarc, I would like to add GGUF support for the OLMo-2 architecture.

I checked and the model is already in transformers, llama.cpp supports the olmo2 arch, and there are GGUF checkpoints on the Hub to test with (for example allenai/OLMo-2-1124-7B-Instruct-GGUF and the small allenai/OLMo-2-0425-1B-GGUF). It is not in ggml.py yet and I could not find an existing PR for it.

I plan to follow the same steps as PR #31175: add the config and tensor mappings, add a tokenizer converter, and add a test. Could I take this one? Thanks!

**Comment by Madhav-000-s** [comment]
Hi @SunMarc! I'd like to take up cohere2 (c4ai-command-a-03-2025 / c4ai-command-r7b-12-2024) from the checklist — following up on the earlier request for it in this thread.

Quick summary of the approach, following the protocol from the issue description and recent arch additions like #44526:

Add the cohere2 entry to GGUF_CONFIG_MAPPING (tensor mapping should come via gguf-py's MODEL_ARCH.COHERE2 name map)
Wire up the tokenizer converter (Cohere GGUFs use BPE, so starting from GGUFGPTConverter)
Add integration tests using a Q2_K quant of Command R7B (small enough for CI), verifying generation and tokenizer parity against the original checkpoint
Will open a PR shortly!

**Comment by gideonphilip** [comment]
Hi @SunMarc 👋

I'd like to claim **DeepSeek-V3** from the checklist and add GGUF loading support for it.

A bit of homework before starting:
- `deepseek_v3` is already integrated in `transformers`, so the model prerequisite is met.
- I don't see DeepSeek-V3 in `GGUF_CONFIG_MAPPING` / `GGUF_TENSOR_MAPPING` yet, and I couldn't find an open PR adding it (the last mention in this thread was ~May 2025 and didn't lead to a PR), so it looks unclaimed.
- There are GGUF checkpoints on the Hub to validate against (e.g. the `unsloth/DeepSeek-V3-GGUF` conversions), so I can add a loading test.

Planned approach, following the protocol in the issue description and prior additions (e.g. #31175, and the Qwen2MoE PR #33264 as an MoE reference):
1. Add `deepseek_v3` entries to `GGUF_CONFIG_MAPPING` and `GGUF_TENSOR_MAPPING` in `integrations/ggml.py`, handling the MoE expert tensors and the MLA (multi-head latent attention) projections, which differ from the dense-model mappings.
2. Handle the config/tokenizer conversion as needed.
3. Add a GGUF loading test.

Does this sound right, and is the architecture free to take? Happy to adjust based on your guidance before opening a PR. Thanks!

**Comment by vibhorxpandey** [comment]
Hi @SunMarc, I'd like to add GGUF support for Phi-2. The phi model is already in transformers and llama.cpp uses the phi2 architecture. I have a working implementation that maps the phi2 config, registers the GPT2 BPE tokenizer converter, splits the fused attn-qkv projection into separate q/k/v, and derives partial rotary factor from rope.dimension_count. I'll open a PR shortly.

**Comment by yun520-1** [comment]
The GGUF support expansion is valuable, but the real bottleneck is **quantization fidelity at the operator level**, not just file loading. GGUF converts weights to a target bitwidth, but if a kernel was written for FP16 dequantization, the INT4 path is often hand-waved with `dequantize_to_fp32` inside the hot loop.

What we'd suggest tracking per architecture:
- which operators have native INT4/INT8 kernels vs. dequantize-then-compute fallbacks
- the accuracy delta between native and fallback paths on standard benchmarks
- a CI matrix that runs a 10B-parameter model through each kernel path on every PR

This turns 'GGUF support' from a binary flag into a measurable property.

**Comment by YeonwooSung** [comment]
@SunMarc checking what's left before I start anything.

I will not open a PR for llama4, DeepSeek-V3, or Command-A (#45546, #47421, #47052). glm4moe (#46718) and phi-2 (#47400) are taken too. Local `GGUF_CONFIG_MAPPING` already has gemma4.

After #47779 the in-flight work has been on `src/transformers/integrations/gguf/` (#48529, #48656, #48660). Should a community PR use that path, or the legacy mappings in `ggml.py`?

If a community PR is still useful on the new path, is **OLMo-2** (`olmo2` / llama.cpp `OLMO2`) acceptable? @0xpinara claimed it here on 2026-06-30 but I cannot find a follow-up PR — still on it? If that claim is live I will leave it.

I will wait for an ACK on both the architecture and the integration path.
